packages/ai-jsx/src/core/opentelemetry.ts

```diff
@@ -31,24 +30,37 @@ interface MemoizedIdAndOwner {
   isOwner: boolean;
 }
 
+// Indicates whether we're currently tracing _any_ component so that we can skip
+// tracing non-root synchronous components.
+const isTracingComponent = createContextKey('Whether the current async context is tracing a component');
+
 // Indicates whether the current async context is responsible for tracing the memoized subtree.
 const activeMemoizedIdStorage = createContextKey('The active MemoizedIdAndOwner');
 
 function spanAttributesForElement(element: Element<any>): opentelemetry.Attributes {
   return { 'ai.jsx.tag': element.tag.name, 'ai.jsx.tree': debug(element, true) };
 }
 
-const getEncoder = _.once(() => getEncoding('cl100k_base'));
-
 export function openTelemetryStreamRenderer(streamRenderer: StreamRenderer): StreamRenderer {
+  const traceEveryComponent = Boolean(getEnvVar('AIJSX_OPENTELEMETRY_TRACE_ALL', false));
   const tracer = opentelemetry.trace.getTracer('ai.jsx');
 
-  return (renderContext, renderable, shouldStop, appendOnly) => {
+  const wrappedRender: StreamRenderer = (renderContext, renderable, shouldStop, appendOnly) => {
     if (!isElement(renderable)) {
       return streamRenderer(renderContext, renderable, shouldStop, appendOnly);
     }
 
-    const activeContext = context.active();
+    let activeContext = context.active();
+    if (!traceEveryComponent && renderable.tag.constructor === Function && activeContext.getValue(isTracingComponent)) {
```

Comment:
Hmm. This seems like a clever heuristic that happens to eliminate most of the AI.JSX internal components, but I think the false positives of "userland components that also get eliminated" will be pretty frustrating ("where are my components?"). In the Sidekick Fantasy Football, for instance, I had [a synchronous user-land component](https://github.com/fixie-ai/sidekick-fantasy-football/blob/e74b07cda8ac953abb1c5fa872da9bc026b960b1/app/api/chat/stats.tsx). 

And the "non-async functions which return promises" thing also seems like a footgun.

What I would suggest instead, and I recognize this is annoyingly boilerplatey, is to explicitly mark every component with some sort of flag:

```ts
function SwitchContext() { }
SwitchContext.aiJSXInternal = true;

// or
SwitchContext.logInTelemetry = true;

// or
SwitchContext.telemetryLogLevel = 'trace';
```

This way, by default, every userland component shows up in the trace, which I think is what we want. And the latter two options I gave would let the user opt out of that as well.

Reply:
Yeah I started looking at this from exactly that angle but convinced myself that the juice wasn't going to be worth the squeeze vs. the async heuristic.

One long-running related thought I've had is whether "structural components" should be made explicitly inert with `string` or `Symbol` tags (e.g. `<AssistantMessage>` might become something like `<assistant>`). This might allow for:
 - natural distinction between no-op components and meaningful components for things like OTel
 - replace partial rendering with something more constrained/easier to reason about
 - avoid the "oops the tree was flattened" footgun
 - native tree serialization

(Don't worry, that's still just an idea at this point.)

Reply:
Oh, I meant to also add, re: the fantasy football example:
- I think the contents of such components are mainly useful when shown as part of the prompt (which is logged both as JSX and OpenAI request object)
- Such components are free to log whatever they want :)
- There's a cheesy but trivial workaround to opt into tracing, which is simply to make the function `async` (even if it doesn't use `await`)